src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,594 @@
+            elif _type == 'loguniform':
+                cs.add_hyperparameter(CSH.UniformFloatHyperparameter(
+                    var, lower=search_space[var]["_value"][0], upper=search_space[var]["_value"][1], 
+                    log=True))
+            elif _type == 'qloguniform':
+                cs.add_hyperparameter(CSH.UniformFloatHyperparameter(
+                    var, lower=search_space[var]["_value"][0], upper=search_space[var]["_value"][1], 
+                    q=search_space[var]["_value"][2], log=True))
+            elif _type == 'normal':
+                cs.add_hyperparameter(CSH.NormalFloatHyperparameter(
+                    var, mu=search_space[var]["_value"][1], sigma=search_space[var]["_value"][2]))
+            elif _type == 'qnormal':
+                cs.add_hyperparameter(CSH.NormalFloatHyperparameter(
+                    var, mu=search_space[var]["_value"][1], sigma=search_space[var]["_value"][2], 
+                    q=search_space[var]["_value"][3]))
+            elif _type == 'lognormal':
+                cs.add_hyperparameter(CSH.NormalFloatHyperparameter(
+                    var, mu=search_space[var]["_value"][1], sigma=search_space[var]["_value"][2], 
+                    log=True))
+            elif _type == 'qlognormal':
+                cs.add_hyperparameter(CSH.NormalFloatHyperparameter(
+                    var, mu=search_space[var]["_value"][1], sigma=search_space[var]["_value"][2], 
+                    q=search_space[var]["_value"][3], log=True))
+            else:
+                raise ValueError(
+                    'unrecognized type in search_space, type is {}'.format(_type))
+
+        self.search_space = cs
+        return True
+
+    def handle_trial_end(self, data):
+        """receive the information of trial end and generate next configuaration.
+
+        Parameters
+        ----------
+        data: dict()
+            it has three keys: trial_job_id, event, hyper_params
+            trial_job_id: the id generated by training service
+            event: the job's state
+            hyper_params: the hyperparameters (a string) generated and returned by tuner
+        """
+        logger.debug('Tuner handle trial end, result is %s', data)
+
+        hyper_params = json_tricks.loads(data['hyper_params'])
+        s, i, _ = hyper_params['parameter_id'].split('_')
+        hyper_configs = self.brackets[int(s)].inform_trial_end(int(i))
+
+        if hyper_configs is not None:
+            logger.debug(
+                'bracket %s next round %s, hyper_configs: %s', s, i, hyper_configs)
+            self.generated_hyper_configs = self.generated_hyper_configs + hyper_configs
+            for _ in range(self.credit):
+                self._request_one_trial_job()
+
+        # Finish this bracket and generate a new bracket
+        if self.brackets[int(s)].no_more_trial:
+            self.curr_s -= 1
+            self.generate_new_bracket()
+        return True
```

Comment:
No docstring to explain it...  Also seems we don't use the return value.

Reply:
This function doesn't have parameter and returns.

Reply:
Done.